opentitan: Add ROM_EXT-compatible ePMP setup.

jrvanwhy opened the pull request:
The existing ePMP code was developed and tested in environments that use a test ROM. On the real silicon, we instead use a ROM_EXT, which passes a different ePMP config to Tock than the test ROM. In particular, it:

1. Sometimes configures and locks entries 0 and 1 (depends on which ROM_EXT is in use).
2. Sets up several entries in a manner that works for Tock, but without the lock bit set.
3. Disables the Rule Lock Bypass bit (this breaks the dance the current constructor does with the MMIO region).

This ePMP config is documented at https://opentitan.org/book/sw/device/silicon_creator/rom_ext/doc/si_val.html#epmp.

This PR renames the existing EarlGreyEPMP constructor to `new_test_rom`, and introduces a new constructor for use with a ROM_EXT, called simply `new`. This constructor makes use of the ROM_EXT's existing layout to make ePMP configuration much simpler; it pretty much just locks down the ROM_EXT's config in place.

This has been tested and verified to work on OpenTitan silicon.

lschuermann commented on chips/earlgrey/src/epmp.rs:
```diff
@@ -384,6 +438,116 @@ impl<const HANDOVER_CONFIG_CHECK: bool> EarlGreyEPMP<{ HANDOVER_CONFIG_CHECK },
+        csr::CSR.pmpcfg1.set(0x00000000);
+        csr::CSR.pmpcfg2.set(0x00000000);
+
+        // ---------- PMP machine CSRs configured, lock down the system
+
+        // Finally, unset the rule-lock bypass (RLB) bit. If we don't have a
+        // debug memory region provided, further set machine-mode lockdown (we
+        // can't enable MML and also have a R/W/X region). We also set MMWP for
+        // good measure, but that shouldn't make a difference -- it can't be
+        // cleared anyways as it is a sticky bit.
+        //
+        // Unsetting RLB with at least one locked region will mean that we can't
+        // set it again, thus actually enforcing the region lock bits.
+        //
+        // Set RLB(2) = 0, MMWP(1) = 1, MML(0) = 1
```
From what I understand, the new `ROM_EXT` will already have this exact(?) `mseccfg` CSR value set for us. If that is the case, I'm by no means opposed to keeping this write in to make this explicit, but we might want to update the comment documenting this expectation.

jrvanwhy replied:
The `ROM_EXT` I'm working with actually sets `0x2` (MMWP, but not MML or RLB). I've updated the documentation to just say we're enabling MML.

lschuermann reviewed (approved):
This looks great to me, generally. Just one minor nit with the `mseccfg` documentation.

Also, I think that we should update the constructor of `new_test_rom` to explicitly lock down regions 2 & 3, now that they're unused (in order to be compatible to the regular `new`), like so:

```diff
          // Set the appropriate `pmpcfg0` register value:
          //
          // 0x80 = 0b10000000, for start address of the kernel .text TOR entry
          //        setting L(7) = 1, A(4-3) = OFF, X(2) = 0, W(1) = 0, R(0) = 0
          //
          // 0x8d = 0b10001101, for kernel .text TOR region
          //        setting L(7) = 1, A(4-3) = TOR,   X(2) = 1, W(1) = 0, R(0) = 1
+         //
+         // 0x80 = 0b10000000, for disabled padding regions 2 & 3 (to be
+         //        compatible with the non test-rom constructor)
+         //        setting L(7) = 1, A(4-3) = OFF, X(2) = 0, W(1) = 0, R(0) = 0
-         csr::CSR.pmpcfg0.set(0x00008d80);
+         csr::CSR.pmpcfg0.set(0x80808d80);
```

lschuermann commented:
We'll probably also want to update the in-tree board to use the `new_test_rom` constructor (in essence not changing it's ePMP instantiation, except for having one less userspace region available), at least until we update to a bitstream where the new, locked-down `EXT_ROM` ships by default.

jrvanwhy commented:
> Also, I think that we should update the constructor of `new_test_rom` to explicitly lock down regions 2 & 3, now that they're unused (in order to be compatible to the regular `new`), like so:

Done.

> We'll probably also want to update the in-tree board to use the `new_test_rom` constructor (in essence not changing it's ePMP instantiation, except for having one less userspace region available), at least until we update to a bitstream where the new, locked-down `EXT_ROM` ships by default.

It currently uses `new_debug`, so I decided to leave it as-is.

jrvanwhy commented:
> Question: does this (and if so how) impact the qemu virtualized opentitan?

I renamed `EarlGreyEPMP::new` to `EarlGreyEPMP::new_test_rom`; it should still work on QEMU as-is. I really down that the new `EarlGreyEPMP::new` function I added in this PR will work on QEMU.

The only change to `EarlGreyEPMP::new_test_rom` is to lock out one of the regions that was previously used for userspace (which is to give it the same functionality in practice as `EarlGreyEPMP::new`).

lschuermann commented:
@alevy One other insight is that (upstream) QEMU does not faithfully emulate any of OpenTitan's ROMs right now. So on QEMU, the default constructors that we have will fail their pre-configuration checks, unless we pass an option that explicitly disables these (at risk of ending up with a configuration that's not secure or soft-bricks the chip, depending on the exact hardware configuration that we start with).
